### PR TITLE
Add CI action to check that set milestone is the next release

### LINKED ISSUE #7079: Milestone checker should check whether the milestone is the next release
## 🧰 Task

Currently the label and milestone checker checks whether there is *a* milestone set for a PR. It should check whether it is the *next* milestone, because otherwise one could set a future milestone for a PR, and *merge* the PR, and but then the release notes process would not pick it up.

### PR DESCRIPTION
**Pull request by Czaki**
# References and relevant issues

closes #7079

# Description

Our current check only checks that *a* milestone is set, but it could be the wrong milestone — for example, 0.5 instead of 0.5.2. Then, if a PR is merged, it will not be included in the release notes. This update ensures that the set milestone before checks pass is the next release.

### CODE REVIEW

**Comment by jni** [comment]
So cool! Thanks for demoing this in the eurasia community meeting @Czaki. As mentioned in the meeting, one update I would do here is to add the earliest milestone and the tagged milestone to the failure message.

It might be useful to make the message multiple lines and explain what can be done about it: "If this PR can be merged in time for the earlier milestone, changing the milestone to {earlier milestone} will make the check pass. If this PR must wait until {currently set milestone}, remove the ready-to-merge tag to skip this check, and re-add it when all earlier milestones are completed."

Alternatively we could have a docs page that explains the labeling and milestone process and link directly to the section dealing with this action.

Overall though, I really like it.

File: .github/workflows/label_and_milesone_checker.yml
Comment (brisvag) [inline .github/workflows/label_and_milesone_checker.yml]:
```suggestion
            echo "remove the ready-to-merge tag to skip this check,"
```

**Review by brisvag — APPROVED** [review]

File: .github/workflows/label_and_milesone_checker.yml
Comment (jni) [inline .github/workflows/label_and_milesone_checker.yml:55]: I actually don't think that we should `map(select(.dueOn >= now))` because I think it will happen relatively often that we miss a deadline by a couple of days. And that's *precisely* the time when this check becomes super important. 😅 

Of course, the converse problem is that we forget to close a deadline. But, brainwave 🧠 🌊, what if we add a step to the release action that closes the corresponding milestone??? 🤯

**Comment by jni** [comment]
Screenshot 2024-07-11 at 12 04 08 PM

I also note that the button is super green for merging even when the check is skipped. I think it used to be that it was gray, not red, before, if the check had been skipped? 🤔

**Comment by jni** [comment]
Screenshot 2024-07-11 at 12 24 10 PM

This is what I'm talking about, from the docs repo

File: .github/workflows/label_and_milesone_checker.yml
Comment (Czaki) [reply .github/workflows/label_and_milesone_checker.yml:55]: I'm not very familiar with this API. So I prefer to wait with experiments until 0.5.0 is released and closed to check if removing this will not catch closed milestones.

**Comment by jni** [comment]
> The grayed out button is when some action is not finished yet. To implement it, we need to use GitHub app, not GitHub action.

Oooooh, interesting!

**Comment by jni** [comment]
Should I enable it on this repo also? Seems handy (given that the PR is all green right now with no milestone)

**Comment by Czaki** [comment]
> Should I enable it on this repo also? Seems handy (given that the PR is all green right now with no milestone)

But even with this action, the PR will be green when 0.5.1 milestone is set. We may need to develop custom action...

**Comment by jni** [comment]
Actually, here's an idea:
- enable the milestone checker app on this repo
- trigger the label and milestone action any time the milestone changes

**Comment by jni** [comment]
this way, the state is pending, not red, until the milestone is set — by a core dev, and then it's ok for the label and milestone checker to run.

**Comment by Czaki** [comment]
And the whole development of this action file was started because you are unhappy with this GitHub app. We may try to go into this direction or check how hard it will be developing own action.

**Comment by jni** [comment]
Let's see how the combo of current app + this action works out.

File: .github/workflows/label_and_milesone_checker.yml
Comment (jni) [inline .github/workflows/label_and_milesone_checker.yml]:
```suggestion
  check_next_milestone:
    name: Check milestone is next release
```

**Review by jni — CHANGES_REQUESTED** [review]
Love this!

I think we should also rename the earlier job from "Check labels and milestone" to "Check PR has type label" (line 18), and remove the milestone check from that job (which is now redundant) (lines 27-31)

**Comment by jni** [comment]
Hmmm, now actionlint is unhappy even though everything was working fine... 🤦 🤔

**Comment by Czaki** [comment]
> Hmmm, now actionlint is unhappy even though everything was working fine... 🤦 🤔

The core problem is that in `gh api graphql` there is a need to define a function to perform the query. And query language uses `$` to evaluate variable, same as in bash scripts. So shellcheck found `$something` and warn that variable cannot be evaluated in `'` quotes. That is intentional, 

So I disable this rule in script.

**Comment by jni** [comment]
Hmmm... @Czaki any idea why the check is stuck on "expected" now...??? 🤔

**Comment by jni** [comment]
It's weird... Actually both checks have passed...

**Review by jni — APPROVED** [review]

**Comment by jni** [comment]
... Literally cannot merge. wth GH. 😂

**Pull request merged**
